runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,529 @@
+    ReinitializedForRecentBlockhash,
+}
+
+pub type InstalledSchedulerBox = Box<dyn InstalledScheduler<DefaultScheduleExecutionArg>>;
+// somewhat arbitrary new type just to pacify Bank's frozen_abi...
+#[derive(Debug, Default)]
+pub(crate) struct InstalledSchedulerBoxInBank(Option<InstalledSchedulerBox>);
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+use solana_frozen_abi::abi_example::AbiExample;
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl AbiExample for InstalledSchedulerBoxInBank {
+    fn example() -> Self {
+        Self(None)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct SchedulingContext {
+    mode: SchedulingMode,
+    // Intentionally not using Weak<Bank> for performance reasons
+    bank: Arc<Bank>,
+}
+
+impl WithSchedulingMode for SchedulingContext {
+    fn mode(&self) -> SchedulingMode {
+        self.mode
+    }
+}
+
+impl SchedulingContext {
+    pub fn new(mode: SchedulingMode, bank: Arc<Bank>) -> Self {
+        Self { mode, bank }
+    }
+
+    pub fn bank(&self) -> &Arc<Bank> {
+        &self.bank
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.bank().slot()
+    }
+
+    pub fn log_prefix(context: Option<&Self>, scheduler_id: SchedulerId) -> String {
+        const BITS_PER_HEX_DIGIT: usize = 4;
+
+        format!(
+            "sch_{:0width$x}{}: ",
+            scheduler_id,
+            context
+                .map(|c| format!("(slot:{}/mode:{:?})", c.slot(), c.mode()))
+                .unwrap_or_else(|| "(?)".into()),
+            width = SchedulerId::BITS as usize / BITS_PER_HEX_DIGIT,
+        )
+    }
+}
+
+// tiny wrapper to ensure to call schedule_termination() via ::drop() inside
+// BankForks::set_root()'s pruning.
+pub(crate) struct BankWithScheduler(Arc<Bank>);
+
+impl BankWithScheduler {
+    pub(crate) fn new(bank: Arc<Bank>) -> Self {
+        Self(bank)
+    }
+
+    pub(crate) fn bank_cloned(&self) -> Arc<Bank> {
+        self.0.clone()
+    }
+
+    pub(crate) fn bank(&self) -> &Arc<Bank> {
+        &self.0
+    }
+
+    pub(crate) fn into_bank(self) -> Arc<Bank> {
+        let bank = self.bank_cloned();
+        drop(self);
+        bank
+    }
+}
+
+impl Drop for BankWithScheduler {
+    fn drop(&mut self) {
+        self.0.schedule_termination();
+    }
+}
+
+impl Deref for BankWithScheduler {
+    type Target = Bank;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl BankForks {
+    pub fn install_scheduler_pool(&mut self, pool: InstalledSchedulerPoolArc) {
+        info!("Installed new scheduler_pool into bank_forks: {:?}", pool);
+        assert!(self.scheduler_pool.replace(pool).is_none());
+    }
+
+    pub(crate) fn install_scheduler_into_bank(&self, bank: &Arc<Bank>) {
+        if let Some(scheduler_pool) = &self.scheduler_pool {
+            let context = SchedulingContext::new(SchedulingMode::BlockVerification, bank.clone());
+            bank.install_scheduler(scheduler_pool.take_from_pool(context));
+        }
+    }
+}
+
+impl Bank {
```

Comment:
At least for me, I strongly prefer this impl to be inside the bank.rs file. While reviewing non-bank code I found myself seeing `bank.XXX` and then I'd hop over to bank.rs to look for `XXX` and what it was doing only to find it doesn't live there.

Reply:
hmm, i'm seeming opposite direction there: we're actively breaking down the huge `bank.rs` into pieces. and i like it.

```
$ git grep "impl Bank "
runtime/src/bank.rs:impl Bank {
runtime/src/bank/sysvar_cache.rs:impl Bank {
runtime/src/bank/transaction_account_state_info.rs:impl Bank {
```

note that bunch of test is separated out into files already as well.

so, i followed the new fancy trend. :)

Reply:
> to be sure, i don't like the current state of affairs: Bank is a god class without doubt and smells badly. but fixing it altogether isn't easy..

Happy we agree with `Bank` being too big. I am trying to push back to a reasonable degree to avoid making the problem worse, but not I also don't want to sacrifice performance for "clean code"